myst_nb/nb_glue/domain.py

```diff
@@ -0,0 +1,278 @@
+        if align:
+            figure_node["align"] = align
+        self.add_name(figure_node)
+        # note: this is copied directly from sphinx.Figure
+        if self.content:
+            node = nodes.Element()  # anonymous container for parsing
+            self.state.nested_parse(self.content, self.content_offset, node)
+            first_node = node[0]
+            if isinstance(first_node, nodes.paragraph):
+                caption = nodes.caption(first_node.rawsource, "", *first_node.children)
+                caption.source = first_node.source
+                caption.line = first_node.line
+                figure_node += caption
+            elif not (isinstance(first_node, nodes.comment) and len(first_node) == 0):
+                error = self.state_machine.reporter.error(
+                    "Figure caption must be a paragraph or empty comment.",
+                    nodes.literal_block(self.block_text, self.block_text),
+                    line=self.lineno,
+                )
+                return [figure_node, error]
+            if len(node) > 1:
+                figure_node += nodes.legend("", *node[1:])
+        return [figure_node]
+
+
+def paste_any_role(name, rawtext, text, lineno, inliner, options={}, content=[]):
+    """This role will simply add the cell output
+    """
+    path = inliner.document.current_source
+    # Remove line number if we have a notebook because it is unreliable
+    if path.endswith(".ipynb"):
+        lineno = None
+    path = str(Path(path).with_suffix(""))
+    return [PasteInlineNode(text, location=(path, lineno))], []
+
+
+def paste_text_role(name, rawtext, text, lineno, inliner, options={}, content=[]):
+    """This role will be parsed as text, with some formatting fanciness.
+
+    The text can have a final ``:``,
+    whereby everything to the right will be treated as a formatting string, e.g.
+    ``key:.2f``
+    """
+    # First check if we have both key:format in the key
+    parts = text.rsplit(":", 1)
+    if len(parts) == 2:
+        key, formatting = parts
+    else:
+        key = parts[0]
+        formatting = None
+
+    path = inliner.document.current_source
+    # Remove line number if we have a notebook because it is unreliable
+    if path.endswith(".ipynb"):
+        lineno = None
+    path = str(Path(path).with_suffix(""))
+    return [PasteTextNode(key, formatting=formatting, location=(path, lineno))], []
+
+
+class NbGlueDomain(Domain):
+    """A sphinx domain for handling glue data """
+
+    name = "glu"
+    label = "NotebookGlue"
+    # data version, bump this when the format of self.data changes
+    data_version = 0.1
+    # data value for a fresh environment
+    # - cache is the mapping of all keys to outputs
+    # - docmap is the mapping of docnames to the set of keys it contains
+    # TODO storing all the outputs in the cache, could allow it to get very big
+    # we may need to consider storing outputs on disc?
```

Comment:
This should definitely be addressed at some point; if you end up 'glueing' 100s of plots, the cache is going to get very large, and most probably affect the load speed / memory usage of the sphinx environment

Reply:
yeah for sure - maybe as a first step, we can add a check for items with type `image/png` in the notebooks and raise a warning once it gets over some arbitrary-but-high number?

Reply:
Yeh all of that. Since we are already writing all these outputs to disc: https://github.com/ExecutableBookProject/MyST-NB/blob/1f88b44f88ec3051ca2d3cf5994a56980a377d78/myst_nb/parser.py#L151, it would be ideal to just link to those

Reply:
Ah good point I think all we'd need to do is remove the data section from those outputs then

Reply:
Ah yeh that's the one, fixed in c0a530c